Eventually sometimes only checks a single time despite short  scheduled intervals

**Which version of Kotest are you using**
4.4.1

```java
java.lang.AssertionError: Eventually block failed after 200ms; attempted 1 time(s); FibonacciInterval(base=20.0ms, offset=0, max=120m) delay between attempts
```
I regularly encountered this now. Per se it isn't a big problem if the intervals aren't followed exactly, if the tests keep my processors busy it doesn't really matter. However, the eventually should definitely do at least a second check at the end of its interval.

--- a/gradle/wrapper/gradle-wrapper.properties
+++ b/gradle/wrapper/gradle-wrapper.properties
@@ -3,4 +3,4 @@ distributionBase=GRADLE_USER_HOME
 distributionPath=wrapper/dists
 zipStoreBase=GRADLE_USER_HOME
 zipStorePath=wrapper/dists
-distributionUrl=https\://services.gradle.org/distributions/gradle-6.8-all.zip
+distributionUrl=https\://services.gradle.org/distributions/gradle-6.8.2-all.zip

--- a/kotest-assertions/kotest-assertions-core/src/jvmTest/kotlin/com/sksamuel/kotest/assertions/timing/EventuallyTest.kt
+++ b/kotest-assertions/kotest-assertions-core/src/jvmTest/kotlin/com/sksamuel/kotest/assertions/timing/EventuallyTest.kt
@@ -21,11 +21,16 @@ import kotlinx.coroutines.delay
 import java.io.FileNotFoundException
 import java.io.IOException
 import java.util.concurrent.CountDownLatch
+import java.util.concurrent.Executors
 import java.util.concurrent.TimeUnit
+import java.util.concurrent.atomic.AtomicInteger
 import kotlin.time.TimeSource
 import kotlin.time.days
 import kotlin.time.milliseconds
 import kotlin.time.seconds
+import kotlinx.coroutines.asCoroutineDispatcher
+import kotlinx.coroutines.async
+import kotlinx.coroutines.withContext
 
 class EventuallyTest : WordSpec() {
 
@@ -36,7 +41,7 @@ class EventuallyTest : WordSpec() {
                System.currentTimeMillis()
             }
          }
-         "pass tests that completed within the time allowed"  {
+         "pass tests that completed within the time allowed" {
             val end = System.currentTimeMillis() + 250
             eventually(1.seconds) {
                if (System.currentTimeMillis() < end)
@@ -130,6 +135,37 @@ class EventuallyTest : WordSpec() {
                count += 1
             }
             count.shouldBeLessThan(6)
+         }
+         "do one final iteration if we never executed before interval expired" {
+            val dispatcher = Executors.newSingleThreadExecutor().asCoroutineDispatcher()
+            async(dispatcher) {
+               Thread.sleep(2000)
+            }
+            val counter = AtomicInteger(0)
+            withContext(dispatcher) {
+               // we won't be able to run in here
+               eventually(1.seconds, 100.milliseconds) {
+                  counter.incrementAndGet()
+               }
+            }
+            counter.get().shouldBe(1)
+         }
+         "do one final iteration if we only executed once and the last delay > interval" {
+            val dispatcher = Executors.newSingleThreadExecutor().asCoroutineDispatcher()
+            // this will start immediately, free the dispatcher to allow eventually to run once, then block the thread
+            async(dispatcher) {
+               delay(100.milliseconds)
+               Thread.sleep(500)
+            }
+            val counter = AtomicInteger(0)
+            withContext(dispatcher) {
+               // this will execute once immediately, then the earlier async will steal the thread
+               // and then since the delay has been > interval and times == 1, we will execute once more
+               eventually(250.milliseconds, 25.milliseconds) {
+                  counter.incrementAndGet() shouldBe 2
+               }
+            }
+            counter.get().shouldBe(2)
          }
          "handle shouldNotBeNull" {
             val mark = TimeSource.Monotonic.markNow()

--- a/kotest-assertions/kotest-assertions-shared/src/commonMain/kotlin/io/kotest/assertions/timing/eventually.kt
+++ b/kotest-assertions/kotest-assertions-shared/src/commonMain/kotlin/io/kotest/assertions/timing/eventually.kt
@@ -77,6 +77,7 @@ suspend fun <T> eventually(
    listener: EventuallyListener<T> = EventuallyListener { },
    f: suspend () -> T,
 ): T {
+
    val start = TimeSource.Monotonic.markNow()
    val end = start.plus(config.duration)
    var times = 0
@@ -86,7 +87,16 @@ suspend fun <T> eventually(
    val originalAssertionMode = errorCollector.getCollectionMode()
    errorCollector.setCollectionMode(ErrorCollectionMode.Hard)
 
-   while (end.hasNotPassedNow() && times < config.retries) {
+   var lastDelayPeriod = Duration.ZERO
+   var lastInterval = Duration.ZERO
+
+   fun attemptsLeft() = end.hasNotPassedNow() && times < config.retries
+
+   // if we only executed once, and the last delay was > last interval, we didn't get a chance to run again
+   // so we run once more before exiting
+   fun isLongWait() = times == 1 && lastDelayPeriod > lastInterval
+
+   while (attemptsLeft() || isLongWait()) {
       try {
          val result = f()
          listener.onEval(EventuallyState(result, start, end, times, firstError, lastError))
@@ -108,7 +118,12 @@ suspend fun <T> eventually(
          }
       }
       times++
-      delay(config.interval.next(times))
+      lastInterval = config.interval.next(times)
+      println("sleeping for $lastInterval")
+      val delayMark = TimeSource.Monotonic.markNow()
+      delay(lastInterval)
+      lastDelayPeriod = delayMark.elapsedNow()
+      println("actual sleep $lastDelayPeriod")
    }
 
    errorCollector.setCollectionMode(originalAssertionMode)

